### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): universally injective morphisms

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
Regarding the large-import, it might be worth investigating whether the imports of `PullbackCarrier.lean` (the newly imported file) can be reduced. But on the other hand `PullbackCarrier.lean` is definitely a basic file for algebraic geometry so I don't think it's a problem to import it freely.

File: Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.PullbackCarrier
+import Mathlib.Topology.LocalAtTarget
+
+/-!
+# Universally closed morphism
+
+A morphism of schemes `f : X ⟶ Y` is universally closed if `X ×[Y] Y' ⟶ Y'` is injective
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean:12]:
```suggestion
# Universally injective morphism

A morphism of schemes `f : X ⟶ Y` is universally injective if `X ×[Y] Y' ⟶ Y'` is injective
```

File: Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+for all base change `Y' ⟶ Y`. This is equivalent to the diagonal morphism being surjective
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean:13]:
```suggestion
for all base changes `Y' ⟶ Y`. This is equivalent to the diagonal morphism being surjective
```
I would even prefer `for all morphisms ...`

File: Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+open CategoryTheory.MorphismProperty Function
+
+/--
+A morphism of schemes `f : X ⟶ Y` is universally injective if the base change `X ×[Y] Y' ⟶ Y'`
+along any morphism `Y' ⟶ Y` is injective (on points).
+-/
+@[mk_iff]
+class UniversallyInjective (f : X ⟶ Y) : Prop where
+  out : universally (topologically (Injective ·)) f
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean:44]: Is there a rule on fields named `out`? In this case, the class name is descriptive enough I suppose.

File: Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+theorem universallyInjective_respectsIso : RespectsIso @UniversallyInjective :=
+  universallyInjective_eq_diagonal.symm ▸ inferInstance
+
+instance universallyInjective_isStableUnderBaseChange :
+    IsStableUnderBaseChange @UniversallyInjective :=
+  universallyInjective_eq_diagonal.symm ▸ inferInstance
+
+instance isInjectiveMap_isStableUnderComposition :
+    IsStableUnderComposition (topologically (Injective ·)) where
+  comp_mem _ _ hf hg := hg.comp hf
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean:83]: This should go to `Mathlib.AlgebraicGeometry.Morphisms.UnderlyingMap`.

File: Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    UniversallyInjective f ↔ Surjective (pullback.diagonal f) := by
+  rw [universallyInjective_eq_diagonal]; rfl
+
+instance (priority := 900) [Mono f] : UniversallyInjective f :=
+  have := (pullback.isIso_diagonal_iff f).mpr inferInstance
+  (UniversallyInjective.iff_diagonal f).mpr inferInstance
+
+theorem universallyInjective_respectsIso : RespectsIso @UniversallyInjective :=
+  universallyInjective_eq_diagonal.symm ▸ inferInstance
+
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean:76]:
```suggestion
theorem UniversallyInjective.respectsIso : RespectsIso @UniversallyInjective :=
  universallyInjective_eq_diagonal.symm ▸ inferInstance
```
Should we converge to this naming scheme for these instances?

File: Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    IsStableUnderComposition @UniversallyInjective :=
+  universallyInjective_eq ▸ inferInstance
+
+instance : MorphismProperty.IsMultiplicative @UniversallyInjective where
+  id_mem _ := inferInstance
+
+instance universallyInjective_isLocalAtTarget : IsLocalAtTarget @UniversallyInjective :=
+  universallyInjective_eq_diagonal.symm ▸ inferInstance
+
+end AlgebraicGeometry
```
Comment (chrisflav) [inline Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean:95]: Could you add a lemma `injective_of_universallyInjective`?

File: Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean (hunk 3 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean]: I've renamed it.

**Review by chrisflav — APPROVED** [review]
I agree that we should treat `PullbackCarrier` as a basic file that may be imported freely. It is still worth to try reducing the imports, but I think it is fine to do this in a different PR.

**Comment by chrisflav** [comment]
Thanks

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Comment by TwoFX** [comment]
This PR uses a `range_iff_surjective` lemma which was renamed in the meantime.
bors r-

**Comment by TwoFX** [comment]
bors d+

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**